framework/scripts/generate_test_list.py

```diff
@@ -74,32 +85,83 @@ def main():
+                f.write(f'static const std::array<struct test*, {num_tests}> test_list_{file_id}{{')
+                test_list = [ f'\n    &_test_{test}' for test in tests ]
+                f.write(','.join(test_list))
+                f.write('\n};\n')
+            f.write('\n')
+
+        # build all defined lists
+        for name, path in test_list_files.items():
+            file_id = files[path]
+            f.write(f'static const struct TestList {name}_test_list{{ {"cpu_" + name if name != "default" else 0}, "{name}", test_list_{file_id} }};\n')
+        # create default "null" list if not defined
+        if default_name is None:
+            f.write(f'static const struct TestList default_test_list{{ 0, "default", std::nullopt }};\n')
+        f.write('\n')
+
+        # function to check if the list is selected, applied to non-default lists only
+        if other_names:
+            f.write('static bool test_list_matches(const TestList* test_list, const char* name)\n')
+            f.write('{\n')
+            f.write('#ifdef SANDSTONE\n')
+            f.write('    if ((name == nullptr) || (strcmp(name, "auto") == 0)) {\n')
+            f.write('        return (test_list->features & cpu_features) == test_list->features;\n')
+            f.write('    }\n')
+            f.write('#endif\n')
+            f.write('    return (name != nullptr) && (strcmp(name, test_list->name) == 0);\n')
+            f.write('}\n\n')
+
+        # selector function
+        f.write('const TestList& select_test_list(const char *name)\n')
+        f.write('{\n')
+        for name in test_list_files.keys():
+            if default_name is None or name != default_name:
+                f.write(f'    if (test_list_matches(&{name}_test_list, name)) return {name}_test_list;\n')
+
+        # emit warning if neither list matches
         if default_name is not None:
-            f.write(f'    if (!test_list_name) return {default_name}_test_list;\n')
-        for name in test_lists.keys():
-            f.write(f'    if  (!strcmp(test_list_name, "{name}")) return {name}_test_list;\n')
-        f.write('    return nullopt;\n')
+            f.write('    if (name && (strcmp(name, "default") != 0)) {\n')
+            f.write('        logging_printf(LOG_LEVEL_QUIET, "# ERROR: list for %s not found. Using default\\n", name);\n')
+            f.write('    }\n')
+        f.write('    return default_test_list;\n')
         f.write('}\n')
 
     # generate .h file
     with open(h_file, 'w') as f:
+        f.write("// **This is generated file, do not edit**\n\n")
+        f.write('#include "sandstone.h"\n')
         f.write('#include <optional>\n')
-        f.write('#include <span>\n')
-        f.write('#include "sandstone.h"\n\n')
-        f.write('std::optional<const std::span<struct test * const>> get_test_list(const char *);\n')
+        f.write('#include <span>\n\n')
+
+        # per-gen list definition
+        f.write('typedef struct TestList {\n')
```

Comment:
This does not need to be here. It should go together with the declaration of `select_test_list()` and implementation of `get_test_list()` to `.cpp`. (`get_test_list()` does not need to be inline, as it does not optimize anything).

Reply:
I've considered moving the header from generated code to the source (what seems to be obvious step, and currently one compile time dependency was removed). Decided to keep both files always generated to reduce number of "visible" changes.
"static inline" is to avoid putting this function in the library interface (regardless optimization level). If it is not used, the body is not even compiled.
select_test_list() has been requested by Thiago ("for future use", ask him for the details).

Reply:
I am saying this should go to the generated .cpp, not to the fixed code. It should not be exposed at all.